HSEARCH-2366 Add an Elasticsearch mapping export tool (#3437)
hibernate/hibernate-search: documentation/src/main/asciidoc/reference/schema.asciidoc

|[[mapper-orm-schema-management-manager-drop-and-create]]`dropAndCreate()`
|Drops existing indexes and re-creates them and their schema.
|<<schema-management-concepts-drop-loses-data,All indexed data will be lost>>.
|[[mapper-orm-schema-management-manager-export-schema]]`exportSchema(Path)`
|Exports the schema into a provided target directory. The output is backend specific.
In case of the <<backend-elasticsearch,Elasticsearch backend>> - json files representing index mapping
and some additional information on the index will be generated in a directory tree grouped by backends and indexed types.
While for the <<backend-lucene,Lucene backend>>, only a file tree will be generated representing the backends and indexed types,
as this backend does not support the schema exporting.

> |This API is still incubating and might change in the future.

Reviewer: I'd move this to a dedicated section, be it only to add an example?

Author: I'll move it once you take a look at the new API and we finalize it 👍🏻😄

The only workaround is to avoid the schema update on AWS.
It should be avoided in production environments regardless:
see <<schema-management-concepts-update-failure>>.

[[schema-management-export]]
== Exporting the schema

[[schema-management-export-filesystem]]
=== Exporting the schema to a set of files

[WARNING]
====
The schema export is backend specific.
Only the <<backend-elasticsearch,Elasticsearch backend>> provides the information required to create an index.
The <<backend-lucene,Lucene backend>> does not provide any actual schema information beyond the list of indexes.
====

> Hibernate Search ships with a built-in schema collector that writes the content of a schema export to a disk.
> All indexes represented by a `SearchSchemaManager` will be written to their own directory as an `index.json` file containing
> information on the index. The output is backend specific.

Reviewer: The file name is obsolete... But this needs to be reworded anyway as we didn't explain schema collectors at this point, and the user actually doesn't need to know we use a schema collector under the hood:

```suggestion
The <<mapper-orm-schema-management-manager,schema manager>> provides a way to export schemas to the filsystem. The output is backend-specific.

[NOTE]
====
Schema exports are constructed based on the mapping information and configurations (e.g. such as the backend version).
Resulting exports are not compared to or validated against the actual backend schema.
====
```

> The file tree structure of an export is shown below:
> [source]
> ----
> # For a default backend the index schema will be written into:
> .../backend/indexes/<index name>/index.json
> # For additional named backends:
> .../backends/<name of a particular backend>/indexes/<index name>/index.json
> ----

Reviewer: This is out of date.

```suggestion
For Elasticsearch, the files provide the necessary information to create indexes (along with their settings and mapping). The file tree structure of an export is shown below:
[source]
----
# For the default backend the index schema will be written to:
.../backend/indexes/<index name>/create-index.json
.../backend/indexes/<index name>/create-index-query-params.json
# For additional named backends:
.../backends/<name of a particular backend>/indexes/<index name>/create-index.json
.../backends/<name of a particular backend>/indexes/<index name>/create-index-query-params.json
----
```

.Exporting the schema to the filesystem
====
[source, JAVA, indent=0]
----
include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/schema/management/HibernateOrmSchemaManagerIT.java[tags=schema-export]
----
<1> Retrieve the `SearchSchemaManager` from a `SearchSession`.
<2> Export the schema to a target directory.
====

[[schema-management-export-custom]]
=== Exporting to a custom collector

<<mapper-orm-schema-management-manager,Search schema managers>> allow to walk through the schema exports based on the data such managers contains.

> To do so a `SearchSchemaCollector` must be implemented and passed to the schema manager's `exportSchema(..)` method.

Reviewer: ```suggestion
To do so a `SearchSchemaCollector` must be implemented and passed to the schema manager's `exportExpectedSchema(..)` method.
```

> Any schema exports are constructed based on the mapping information and configurations (e.g. such as the backend version).
> Resulting exports are not compared to or validated against the actual backend schema.

Reviewer: ```suggestion
[NOTE]
====
Schema exports are constructed based on the mapping information and configurations (e.g. such as the backend version).
Resulting exports are not compared to or validated against the actual backend schema.
====
```

To do that check <<mapper-orm-schema-management-manager,other methods>> provided by the `SearchSchemaManager`.

.Exporting to a custom collector
====
[source, JAVA, indent=0]
----
include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/schema/management/HibernateOrmSchemaManagerIT.java[tags=walking-the-schema]
----
<1> Retrieve the `SearchSchemaManager` from a `SearchSession`.
<2> Instantiate and pass the `SearchSchemaCollector` to walk a schema.
<3> Create a name from an index and backend names.
====

[TIP]
====

> To access additional, backend specific, functionality an extension to `IndexSchemaExport` can be applied:

Reviewer: ```suggestion
To access backend-specific functionality, an extension to `SchemaExport` can be applied:
```

=====
[source, JAVA, indent=0]
----
include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/schema/management/ElasticsearchHibernateOrmSchemaManagerIT.java[tags=walking-the-schema-extension]
----

> <1> Extend the `IndexSchemaExport` with the Elasticsearch extension.

Reviewer: ```suggestion
<1> Extend the `SchemaExport` with the Elasticsearch extension.
```

Resulting exports are not compared to or validated against the actual backend schema.
====

For Elasticsearch, the files provide the necessary information to create indexes (along with their settings and mapping).
The file tree structure of an export is shown below:
[source]
----
# For the default backend the index schema will be written to:
.../backend/indexes/<index name>/create-index.json
.../backend/indexes/<index name>/create-index-query-params.json
# For additional named backends:
.../backends/<name of a particular backend>/indexes/<index name>/create-index.json
.../backends/<name of a particular backend>/indexes/<index name>/create-index-query-params.json
----

.Exporting the schema to the filesystem
====
[source, JAVA, indent=0]
----
include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/schema/management/HibernateOrmSchemaManagerIT.java[tags=schema-export]
----
<1> Retrieve the `SearchSchemaManager` from a `SearchSession`.
<2> Export the schema to a target directory.
====

[[schema-management-export-custom]]
=== Exporting to a custom collector

> <<mapper-orm-schema-management-manager,Search schema managers>> allow to walk through the schema exports based on the data such managers contains.

Reviewer: Just a grammar mistake I didn't spot earlier.

```suggestion
<<mapper-orm-schema-management-manager,Search schema managers>> allow walking through the schema exports based on the data such managers contains.
```
